boa/src/syntax/ast/node/operator/unary_op/mod.rs

```diff
@@ -76,13 +76,16 @@ impl Executable for UnaryOp {
             }
             op::UnaryOp::Not => self.target().run(context)?.not(context)?.into(),
             op::UnaryOp::Tilde => {
-                let num_v_a = self.target().run(context)?.to_number(context)?;
-                Value::from(if num_v_a.is_nan() {
-                    -1
+                let num_v_a = self.target().run(context)?.to_numeric_number(context)?;
+                if num_v_a.is_nan() || num_v_a.is_infinite() {
+                    Value::from(-1) // special case for inf or nan
+                } else if let Some(num_bigint) = self.target.run(context)?.as_bigint() {
```

Comment:
We should not call `self.target().run(context)?` multiple times since the expression could be a function with side effects like:
```js
let x = 0;
function y() {
	x += 1;
	return 1;
}

~(y());
```